gateway-spi/src/main/java/org/apache/knox/gateway/util/AuthFilterUtils.java

```diff
@@ -42,212 +44,247 @@
+            unAuthenticatedPaths.add(tokenizer.nextToken());
+        }
+    }
+
+    /**
+     * A method that parses a string (delimiters = ;,) and adds them to the
+     * provided un-authenticated path set.
+     * @param unAuthenticatedPaths
+     * @param list
+     * @param defaultList
+     */
+    public static void addUnauthPaths(final Set<String> unAuthenticatedPaths, final String list, final String defaultList) {
+        /* add default unauthenticated paths list */
+        parseStringThenAdd(unAuthenticatedPaths, defaultList);
+        /* add provided unauthenticated paths list if specified */
+        if (!StringUtils.isBlank(list)) {
+            AuthFilterUtils.parseStringThenAdd(unAuthenticatedPaths, list);
+        }
+    }
+
+    public static void refreshSuperUserGroupsConfiguration(ServletContext context, List<String> initParameterNames, String topologyName, String role) {
+        if (context == null) {
+            throw new IllegalArgumentException("Cannot get proxyuser configuration from NULL context");
+        }
+        refreshSuperUserGroupsConfiguration(context, null, initParameterNames, topologyName, role);
+    }
+
+    public static void refreshSuperUserGroupsConfiguration(FilterConfig filterConfig, List<String> initParameterNames, String topologyName, String role) {
+        if (filterConfig == null) {
+            throw new IllegalArgumentException("Cannot get proxyuser configuration from NULL filter config");
+        }
+        refreshSuperUserGroupsConfiguration(null, filterConfig, initParameterNames, topologyName, role);
+    }
+
+    private static void refreshSuperUserGroupsConfiguration(ServletContext context, FilterConfig filterConfig, List<String> initParameterNames, String topologyName, String role) {
+        final Configuration conf = new Configuration(false);
+        boolean hasProxyGroupParams = false;
+
+        if (initParameterNames != null) {
+            initParameterNames.stream().filter(name -> name.startsWith(PROXYUSER_PREFIX + ".")).forEach(name -> {
+                String value = context == null ? filterConfig.getInitParameter(name) : context.getInitParameter(name);
+                conf.set(name, value);
+            });
+            /* For proxy groups */
+            hasProxyGroupParams = initParameterNames.stream().anyMatch(name -> name.startsWith(PROXYGROUP_PREFIX + "."));
+            if(hasProxyGroupParams) {
+                initParameterNames.stream().filter(name -> name.startsWith(PROXYGROUP_PREFIX + ".")).forEach(name -> {
+                    String value = context == null ? filterConfig.getInitParameter(name) : context.getInitParameter(name);
+                    conf.set(name, value);
+                });
+            }
         }
 
-        @Override
-        public Principal getUserPrincipal() {
-          return remoteRequestUgi::getUserName;
+        ImpersonationProvider impersonationProvider;
+        if(hasProxyGroupParams) {
```

Comment:
Why can't this be our default implementation that supports both user-level and the new group-level impersonation config? It would simplify a lot of things. So, rather than GroupBasedImpersonationProvider, it would be someyhing like KnoxImpersonationProvider (our implementation that also supports group-level config).